[BUG]: test failures with numpy 1.22.1

### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [X] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### Problem description

Tests run against numpy 1.22.1 have a few new failures. It seems that the `str()` method on some types changed output.

```python
============================================================== FAILURES ===============================================================
_____________________________________________________________ test_dtype ______________________________________________________________

simple_dtype = dtype({'names': ['bool_', 'uint_', 'float_', 'ldbl_'], 'formats': ['?', '<u4', '<f4', '<f16'], 'offsets': [0, 4, 8, 16], 'itemsize': 32})

    def test_dtype(simple_dtype):
        from sys import byteorder
    
        e = "<" if byteorder == "little" else ">"
    
>       assert m.print_dtypes() == [
            simple_dtype_fmt(),
            packed_dtype_fmt(),
            "[('a', {}), ('b', {})]".format(simple_dtype_fmt(), packed_dtype_fmt()),
            partial_dtype_fmt(),
            partial_nested_fmt(),
            "[('a', 'S3'), ('b', 'S3')]",
            (
                "{{'names':['a','b','c','d'], "
                + "'formats':[('S4', (3,)),('"
                + e
                + "i4', (2,)),('u1', (3,)),('"
                + e
                + "f4', (4, 2))], "
                + "'offsets':[0,12,20,24], 'itemsize':56}}"
            ).format(e=e),
            "[('e1', '" + e + "i8'), ('e2', 'u1')]",
            "[('x', 'i1'), ('y', '" + e + "u8')]",
            "[('cflt', '" + e + "c8'), ('cdbl', '" + e + "c16')]",
        ]
E       assert ["{'names': [... 'S3')]", ...] == ["{'names':['... 'S3')]", ...]
E         At index 0 diff: "{'names': ['bool_', 'uint_', 'float_', 'ldbl_'], 'formats': ['?', '<u4', '<f4', '<f16'], 'offsets': [0, 4, 8, 16], 'itemsize': 32}" != "{'names':['bool_','uint_','float_','ldbl_'], 'formats':['?','<u4','<f4','<f16'], 'offsets':[0,4,8,16], 'itemsize':32}"
E         Use -v to get the full diff

../../../../tests/test_numpy_dtypes.py:126: AssertionError
____________________________________________________________ test_recarray ____________________________________________________________

simple_dtype = dtype({'names': ['bool_', 'uint_', 'float_', 'ldbl_'], 'formats': ['?', '<u4', '<f4', '<f16'], 'offsets': [0, 4, 8, 16], 'itemsize': 32})
packed_dtype = dtype([('bool_', '?'), ('uint_', '<u4'), ('float_', '<f4'), ('ldbl_', '<f16')])

    def test_recarray(simple_dtype, packed_dtype):
        elements = [(False, 0, 0.0, -0.0), (True, 1, 1.5, -2.5), (False, 2, 3.0, -5.0)]
    
        for func, dtype in [
            (m.create_rec_simple, simple_dtype),
            (m.create_rec_packed, packed_dtype),
        ]:
            arr = func(0)
            assert arr.dtype == dtype
            assert_equal(arr, [], simple_dtype)
            assert_equal(arr, [], packed_dtype)
    
            arr = func(3)
            assert arr.dtype == dtype
            assert_equal(arr, elements, simple_dtype)
            assert_equal(arr, elements, packed_dtype)
    
            # Show what recarray's look like in NumPy.
            assert type(arr[0]) == np.void
            assert type(arr[0].item()) == tuple
    
            if dtype == simple_dtype:
                assert m.print_rec_simple(arr) == [
                    "s:0,0,0,-0",
                    "s:1,1,1.5,-2.5",
                    "s:0,2,3,-5",
                ]
            else:
                assert m.print_rec_packed(arr) == [
                    "p:0,0,0,-0",
                    "p:1,1,1.5,-2.5",
                    "p:0,2,3,-5",
                ]
    
        nested_dtype = np.dtype([("a", simple_dtype), ("b", packed_dtype)])
    
        arr = m.create_rec_nested(0)
        assert arr.dtype == nested_dtype
        assert_equal(arr, [], nested_dtype)
    
        arr = m.create_rec_nested(3)
        assert arr.dtype == nested_dtype
        assert_equal(
            arr,
            [
                ((False, 0, 0.0, -0.0), (True, 1, 1.5, -2.5)),
                ((True, 1, 1.5, -2.5), (False, 2, 3.0, -5.0)),
                ((False, 2, 3.0, -5.0), (True, 3, 4.5, -7.5)),
            ],
            nested_dtype,
        )
        assert m.print_rec_nested(arr) == [
            "n:a=s:0,0,0,-0;b=p:1,1,1.5,-2.5",
            "n:a=s:1,1,1.5,-2.5;b=p:0,2,3,-5",
            "n:a=s:0,2,3,-5;b=p:1,3,4.5,-7.5",
        ]
    
        arr = m.create_rec_partial(3)
>       assert str(arr.dtype) == partial_dtype_fmt()
E       assert "{'names': ['...temsize': 48}" == "{'names':['b...itemsize':48}"
E         - {'names':['bool_','uint_','float_','ldbl_'], 'formats':['?','<u4','<f4','<f16'], 'offsets':[0,4,8,32], 'itemsize':48}
E         + {'names': ['bool_', 'uint_', 'float_', 'ldbl_'], 'formats': ['?', '<u4', '<f4', '<f16'], 'offsets': [0, 4, 8, 32], 'itemsize': 48}
E         ?          +         +        +         +                    +     +      +      +                   +   +  +  +                +

../../../../tests/test_numpy_dtypes.py:241: AssertionError
__________________________________________________________ test_array_array ___________________________________________________________

    def test_array_array():
        from sys import byteorder
    
        e = "<" if byteorder == "little" else ">"
    
        arr = m.create_array_array(3)
>       assert str(arr.dtype) == (
            "{{'names':['a','b','c','d'], "
            + "'formats':[('S4', (3,)),('"
            + e
            + "i4', (2,)),('u1', (3,)),('{e}f4', (4, 2))], "
            + "'offsets':[0,12,20,24], 'itemsize':56}}"
        ).format(e=e)
E       assert "{'names': ['...temsize': 56}" == "{'names':['a...itemsize':56}"
E         - {'names':['a','b','c','d'], 'formats':[('S4', (3,)),('<i4', (2,)),('u1', (3,)),('<f4', (4, 2))], 'offsets':[0,12,20,24], 'itemsize':56}
E         + {'names': ['a', 'b', 'c', 'd'], 'formats': [('S4', (3,)), ('<i4', (2,)), ('u1', (3,)), ('<f4', (4, 2))], 'offsets': [0, 12, 20, 24], 'itemsize': 56}
E         ?          +     +    +    +                +              +              +             +                            +   +   +   +                +

../../../../tests/test_numpy_dtypes.py:288: AssertionError
```

### Reproducible example code

```text
sed -i -e 's:1.21.3:1.22.1:' tests/requirements.txt
nox -e tests-3.9
```

--- a/tests/test_numpy_dtypes.py
+++ b/tests/test_numpy_dtypes.py
@@ -32,8 +32,8 @@ def dt_fmt():
     e = "<" if byteorder == "little" else ">"
     return (
         "{{'names':['bool_','uint_','float_','ldbl_'],"
-        " 'formats':['?','" + e + "u4','" + e + "f4','" + e + "f{}'],"
-        " 'offsets':[0,4,8,{}], 'itemsize':{}}}"
+        "'formats':['?','" + e + "u4','" + e + "f4','" + e + "f{}'],"
+        "'offsets':[0,4,8,{}],'itemsize':{}}}"
     )
 
 
@@ -46,7 +46,7 @@ def simple_dtype_fmt():
 def packed_dtype_fmt():
     from sys import byteorder
 
-    return "[('bool_', '?'), ('uint_', '{e}u4'), ('float_', '{e}f4'), ('ldbl_', '{e}f{}')]".format(
+    return "[('bool_','?'),('uint_','{e}u4'),('float_','{e}f4'),('ldbl_','{e}f{}')]".format(
         np.dtype("longdouble").itemsize, e="<" if byteorder == "little" else ">"
     )
 
@@ -77,7 +77,7 @@ def partial_nested_fmt():
     partial_size = partial_ld_off + ld.itemsize
     partial_end_padding = partial_size % np.dtype("uint64").alignment
     partial_nested_size = partial_nested_off * 2 + partial_size + partial_end_padding
-    return "{{'names':['a'], 'formats':[{}], 'offsets':[{}], 'itemsize':{}}}".format(
+    return "{{'names':['a'],'formats':[{}],'offsets':[{}],'itemsize':{}}}".format(
         partial_dtype_fmt(), partial_nested_off, partial_nested_size
     )
 
@@ -123,25 +123,25 @@ def test_dtype(simple_dtype):
 
     e = "<" if byteorder == "little" else ">"
 
-    assert m.print_dtypes() == [
+    assert [x.replace(' ', '') for x in m.print_dtypes()] == [
         simple_dtype_fmt(),
         packed_dtype_fmt(),
-        "[('a', {}), ('b', {})]".format(simple_dtype_fmt(), packed_dtype_fmt()),
+        "[('a',{}),('b',{})]".format(simple_dtype_fmt(), packed_dtype_fmt()),
         partial_dtype_fmt(),
         partial_nested_fmt(),
-        "[('a', 'S3'), ('b', 'S3')]",
+        "[('a','S3'),('b','S3')]",
         (
-            "{{'names':['a','b','c','d'], "
-            + "'formats':[('S4', (3,)),('"
+            "{{'names':['a','b','c','d'],"
+            + "'formats':[('S4',(3,)),('"
             + e
-            + "i4', (2,)),('u1', (3,)),('"
+            + "i4',(2,)),('u1',(3,)),('"
             + e
-            + "f4', (4, 2))], "
-            + "'offsets':[0,12,20,24], 'itemsize':56}}"
+            + "f4',(4,2))],"
+            + "'offsets':[0,12,20,24],'itemsize':56}}"
         ).format(e=e),
-        "[('e1', '" + e + "i8'), ('e2', 'u1')]",
-        "[('x', 'i1'), ('y', '" + e + "u8')]",
-        "[('cflt', '" + e + "c8'), ('cdbl', '" + e + "c16')]",
+        "[('e1','" + e + "i8'),('e2','u1')]",
+        "[('x','i1'),('y','" + e + "u8')]",
+        "[('cflt','" + e + "c8'),('cdbl','" + e + "c16')]",
     ]
 
     d1 = np.dtype(
@@ -238,7 +238,7 @@ def test_recarray(simple_dtype, packed_dtype):
     ]
 
     arr = m.create_rec_partial(3)
-    assert str(arr.dtype) == partial_dtype_fmt()
+    assert str(arr.dtype).replace(' ', '') == partial_dtype_fmt()
     partial_dtype = arr.dtype
     assert "" not in arr.dtype.fields
     assert partial_dtype.itemsize > simple_dtype.itemsize
@@ -246,7 +246,7 @@ def test_recarray(simple_dtype, packed_dtype):
     assert_equal(arr, elements, packed_dtype)
 
     arr = m.create_rec_partial_nested(3)
-    assert str(arr.dtype) == partial_nested_fmt()
+    assert str(arr.dtype).replace(' ', '') == partial_nested_fmt()
     assert "" not in arr.dtype.fields
     assert "" not in arr.dtype.fields["a"][0].fields
     assert arr.dtype.itemsize > partial_dtype.itemsize
@@ -285,12 +285,12 @@ def test_array_array():
     e = "<" if byteorder == "little" else ">"
 
     arr = m.create_array_array(3)
-    assert str(arr.dtype) == (
-        "{{'names':['a','b','c','d'], "
-        + "'formats':[('S4', (3,)),('"
+    assert str(arr.dtype).replace(' ', '') == (
+        "{{'names':['a','b','c','d'],"
+        + "'formats':[('S4',(3,)),('"
         + e
-        + "i4', (2,)),('u1', (3,)),('{e}f4', (4, 2))], "
-        + "'offsets':[0,12,20,24], 'itemsize':56}}"
+        + "i4',(2,)),('u1',(3,)),('{e}f4',(4,2))],"
+        + "'offsets':[0,12,20,24],'itemsize':56}}"
     ).format(e=e)
     assert m.print_array_array(arr) == [
         "a={{A,B,C,D},{K,L,M,N},{U,V,W,X}},b={0,1},"

--- a/tests/test_numpy_dtypes.py
+++ b/tests/test_numpy_dtypes.py
@@ -123,7 +123,7 @@ def test_dtype(simple_dtype):
 
     e = "<" if byteorder == "little" else ">"
 
-    assert [x.replace(' ', '') for x in m.print_dtypes()] == [
+    assert [x.replace(" ", "") for x in m.print_dtypes()] == [
         simple_dtype_fmt(),
         packed_dtype_fmt(),
         "[('a',{}),('b',{})]".format(simple_dtype_fmt(), packed_dtype_fmt()),
@@ -238,7 +238,7 @@ def test_recarray(simple_dtype, packed_dtype):
     ]
 
     arr = m.create_rec_partial(3)
-    assert str(arr.dtype).replace(' ', '') == partial_dtype_fmt()
+    assert str(arr.dtype).replace(" ", "") == partial_dtype_fmt()
     partial_dtype = arr.dtype
     assert "" not in arr.dtype.fields
     assert partial_dtype.itemsize > simple_dtype.itemsize
@@ -246,7 +246,7 @@ def test_recarray(simple_dtype, packed_dtype):
     assert_equal(arr, elements, packed_dtype)
 
     arr = m.create_rec_partial_nested(3)
-    assert str(arr.dtype).replace(' ', '') == partial_nested_fmt()
+    assert str(arr.dtype).replace(" ", "") == partial_nested_fmt()
     assert "" not in arr.dtype.fields
     assert "" not in arr.dtype.fields["a"][0].fields
     assert arr.dtype.itemsize > partial_dtype.itemsize
@@ -285,7 +285,7 @@ def test_array_array():
     e = "<" if byteorder == "little" else ">"
 
     arr = m.create_array_array(3)
-    assert str(arr.dtype).replace(' ', '') == (
+    assert str(arr.dtype).replace(" ", "") == (
         "{{'names':['a','b','c','d'],"
         + "'formats':[('S4',(3,)),('"
         + e

